torch/_meta_registrations.py

```diff
@@ -505,20 +502,21 @@ def is_fast_path(src, scale, output, padding_idx):
     return output, offset2bag, bag_size, max_indices
 
 
+@torch.library.impl(meta_lib, "diag")
 @torch.library.impl(meta_lib, "diag.out")
-def meta_diag_out(self, dim=0, *, out):
-    assert self.dim() in (1, 2), "matrix or a vector expected"
+@out_wrapper
+def meta_diag(self, dim=0):
+    check(self.dim() in (1, 2), lambda: "matrix or a vector expected")
     if self.dim() == 1:
         sz = self.size(0) + abs(dim)
-        torch._resize_output_(out, (sz, sz), self.device)
-        return out
+        return self.new_empty((sz, sz))
+
+    # case: dim is 2
+    if dim >= 0:
+        sz = min(self.size(0), self.size(1) - dim)
     else:
-        if dim >= 0:
-            sz = min(self.size(0), self.size(1) - dim)
-        else:
-            sz = min(self.size(0) + dim, self.size(1))
-        torch._resize_output_(out, (sz,), self.device)
-        return out
+        sz = min(self.size(0) + dim, self.size(1))
+    return self.new_empty((sz,))
```

Comment:
I'm perplexed why this isn't a decomp, the C++ implementation is

```
Tensor diag(const Tensor& self, int64_t offset) {
  auto ndim = self.dim();
  TORCH_CHECK(ndim == 1 || ndim == 2, "diag(): Supports 1D or 2D tensors. Got ", self.dim(), "D"); 
  if (ndim == 1) {
    return at::diag_embed(self, offset);
  } else {
    // We return a copy of the diagonal
    return at::diagonal_copy(self, offset);
  }
} 
```

Reply:
Even, I am not sure now 😓. But the following PR took care of it

https://github.com/pytorch/pytorch/pull/87180/files#diff-6dffed1ade0ba3e887f9a4eafa3bfcec267ab2365b8adcb91bd391f49b3fd2e3L1022